File: 1022 Sum of Root To Leaf Binary Numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int sum = 0;
    public int sumRootToLeaf(TreeNode root) {
        if (root.left == null && root.right == null) return root.val;
        if (root.left != null) helper(root.left, root.val * 2);
        if (root.right != null) helper(root.right, root.val * 2);
        
        return sum;
    }
    public void helper(TreeNode r, int val) {
        int tmp = val + r.val;
        if (r.left == null && r.right == null) sum += tmp;
        else {
            if (r.left != null) helper(r.left, tmp * 2);
            if (r.right != null) helper(r.right, tmp * 2);
        }
    }
}

// without using global value solution from lee215
/*
public int sumRootToLeaf(TreeNode root) {
    return dfs(root, 0);
}

public int dfs(TreeNode root, int val) {
    if (root == null) return 0;
    val = val * 2 + root.val;
    return root.left == root.right ? val : dfs(root.left, val) + dfs(root.right, val);
}
*/

File: 1022 Sum of Root To Leaf Binary Numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    sum = 0
    def sumRootToLeaf(self, root: TreeNode) -> int:
        if not root.left and not root.right:
            return root.val
        if root.left: self.helper(root.left, root.val * 2)
        if root.right: self.helper(root.right, root.val * 2)
        return self.sum
    
    def helper(self, r, val):
        tmp = val + r.val
        if not r.left and not r.right:
            self.sum += tmp
        else:
            if r.left: self.helper(r.left, tmp * 2)
            if r.right: self.helper(r.right, tmp * 2)
''' Solution with no global value
def sumRootToLeaf(self, root, val=0):
    if not root: return 0
    val = val * 2 + root.val
    if root.left == root.right: return val # This means that both left and right are null
    return self.sumRootToLeaf(root.left, val) + self.sumRootToLeaf(root.right, val)

'''